### PR TITLE
Add HostFiltering to the default web host

### PR DESCRIPTION
**Pull request by Tratcher**
Damian didn't want this code in the template so we're putting it in the default builder instead.
https://github.com/aspnet/templating/pull/374

It will be disabled by default, it only lights up if you define AllowedHosts as a semicolon separated list in config.

### CODE REVIEW

File: src/Microsoft.AspNetCore/WebHost.cs [hunk 1]
```diff
@@ -182,6 +182,16 @@ public static IWebHostBuilder CreateDefaultBuilder(string[] args)
                     logging.AddConsole();
                     logging.AddDebug();
                 })
+                .ConfigureServices((hostingContext, services) =>
+                {
+                    var hosts = hostingContext.Configuration["AllowedHosts"]?.Split(new[] { ';' }, StringSplitOptions.RemoveEmptyEntries);
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore/WebHost.cs]: Hmm hardcoded constants look suspicious. Why isn't `AllowedHosts` defined as a static string somewhere?

File: samples/SampleApp/appsettings.json [hunk 2]
```diff
@@ -1,4 +1,5 @@
 {
+  "AllowedHosts":  "example.com;localhost",
```
Comment (jkotalik) [inline samples/SampleApp/appsettings.json]: nit extra space.

**Comment by JunTaoLuo** [comment]
@jkotalik tests were modified to check for host filtering as part of their test step.

File: src/Microsoft.AspNetCore/WebHost.cs (hunk 1 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore/WebHost.cs]: It's only used in this one spot. There are several of those in this method.

File: src/Microsoft.AspNetCore/Microsoft.AspNetCore.csproj [hunk 3]
```diff
@@ -10,6 +10,7 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.Diagnostics" Version="$(MicrosoftAspNetCoreDiagnosticsPackageVersion)" PrivateAssets="None" />
+    <PackageReference Include="Microsoft.AspNetCore.HostFiltering" Version="$(MicrosoftAspNetCoreHostFilteringPackageVersion)" PrivateAssets="None" />
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore/Microsoft.AspNetCore.csproj]: This makes me kinda sad.

File: src/Microsoft.AspNetCore/WebHost.cs [hunk 4]
```diff
@@ -182,6 +182,16 @@ public static IWebHostBuilder CreateDefaultBuilder(string[] args)
...
+                    services.AddHostFiltering(options =>
+                    {
+                        options.AllowedHosts = (hosts?.Length > 0 ? hosts : new[] { "*" });
+                    });
+
+                    services.AddTransient<IStartupFilter, HostFilteringStartupFilter>();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore/WebHost.cs]: Can you make sure this middleware is optimized for nooping?

Comment (Tratcher) [reply src/Microsoft.AspNetCore/WebHost.cs]: It is. https://github.com/aspnet/BasicMiddleware/blob/2f92fc1d1eaa4033e5b02581da0018dcf8511ea1/src/Microsoft.AspNetCore.HostFiltering/HostFilteringMiddleware.cs#L150

**Pull request merged**